components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +295,245 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
+                return signedJWT.getJWTClaimsSet().getIssuer();
+            //   Obtain client ID from request parameters
+            } else if (clientIdAsReqParam.isPresent()) {
+                return clientIdAsReqParam.get();
+            //   Obtain client ID from the request body
+            } else if (clientIdInContentParamList.isPresent()) {
+                return (String) clientIdInContentParamList.get().get(0);
+            //   Obtain client ID from the authorization header when basic authentication is used
+            } else if (request.getHeader(authHeader) != null) {
+                String authorizationHeader = request.getHeader(authHeader);
+                if (authorizationHeader.split(" ").length == 2) {
+                    if (request.getHeader(authHeader).split(" ")[0].equals("Basic")) {
+                        String authToken = request.getHeader(authHeader).split(" ")[1];
+                        byte[] decodedBytes = Base64.getUrlDecoder().decode(authToken.getBytes(StandardCharsets.UTF_8));
+                        String decodedAuthToken = new String(decodedBytes, StandardCharsets.UTF_8);
+                        if (decodedAuthToken.split(":").length == 2) {
+                            return decodedAuthToken.split(":")[0];
+                        } else {
+                            throw new OAuthClientAuthnException(basicAuthErrorMessage, OAuth2ErrorCodes.INVALID_CLIENT);
+                        }
+                    } else {
+                        return null;
+                    }
+                } else {
+                    throw new OAuthClientAuthnException(basicAuthErrorMessage, OAuth2ErrorCodes.INVALID_CLIENT);
+                }
+            } else {
+                throw new OAuthClientAuthnException("Unable to find client id in the request",
+                        OAuth2ErrorCodes.INVALID_CLIENT);
+            }
+        } catch (ParseException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the signed assertion",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+    }
+
+    /**
+     * Validate whether the request follows MTLS client authentication.
+     *
+     * @param request               Http servlet request.
+     * @param contentParam          Map of request body params.
+     * @return Whether the request follows MTLS client authentication.
+     * @throws OAuthClientAuthnException
+     */
+    private boolean isMTLSAuthentication(HttpServletRequest request, Map<String, List> contentParam)
+            throws OAuthClientAuthnException {
+
+        String certificate = Optional.ofNullable(IdentityUtil.getProperty(OAuthConstants.MTLS_AUTH_HEADER))
+                .orElse("CONFIG_NOT_FOUND");
+
+        String oauthClientID =  request.getParameter(OAuth.OAUTH_CLIENT_ID);
+        if (StringUtils.isBlank(OAuth.OAUTH_CLIENT_ID)) {
+            oauthClientID = (String) contentParam.get(OAuth.OAUTH_CLIENT_ID).get(0);
```

Comment:
Possible ArrayIndexOutOfBoundsException. Validate the length.

Reply:
Fixed in 36cde69f01a0865dc525ff8c963f66e0af53cf89